### PR TITLE
[tt-train] Fixing Linear and Sequential Scheduler Step 0

### PR DESCRIPTION
**Pull request by awliu-TT**
### PR Category
Bug Fix

### Summary
Currently, the linear and sequential LR schedulers do not apply the `start_factor` at construction time, unlike in PyTorch. Thus, the first optimizer step will result in weight updates using the full base unreduced LR. This PR fixes this behavior to align it with PyTorch. As part of the change, a `get_initial_lr()` method is added to the optimizers to mirror PyTorch's `param_group["initial_lr"]` that stores the base LR at construction.

Note some key contract changes:

- Every optimizer now requires `initial_lr` in its state dict like PyTorch has, so this parameter is now required to be in any loaded checkpoint.
- The `MuonWithAdamW` Optimizer is now being rejected from being used with stateful LR schedulers due to https://github.com/tenstorrent/tt-metal/issues/56027.

### Results
When running the C++ nanogpt trainer, we see that the step 0 LR spike is now gone:

image

And here is the new loss over 500 training steps:

image

### CI Results
- [x] [Sanity Tests](https://github.com/tenstorrent/tt-metal/actions/runs/34914924859)
- [x] [Nightly L2](https://github.com/tenstorrent/tt-metal/actions/runs/34914965002) - [main results](https://github.com/tenstorrent/tt-metal/actions/runs/30073306888)
- [x] [(Runtime) Sanity Tests](https://github.com/tenstorrent/tt-metal/actions/runs/34914994188)
- [x] New/Existing unit tests for this change
### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:awliu/scheduler_factor_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:awliu/scheduler_factor_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:awliu/scheduler_factor_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:awliu/scheduler_factor_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:awliu/scheduler_factor_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:awliu/scheduler_factor_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:awliu/scheduler_factor_fix)

### CODE REVIEW

File: tt-train/sources/ttml/ttml/common/schedulers.py [hunk 1]
```diff
@@ -240,6 +276,12 @@ def __init__(self, optimizer, schedulers, milestones):
         self._milestones = list(milestones)
         self._current_scheduler_index = 0
         self._current_step_in_scheduler = 0
+        # The children were constructed back-to-back on the same optimizer, so
+        # the optimizer currently holds the LAST child's construction-time LR.
+        # Only the first child is active; restore its initial LR. Mirrors
+        # PyTorch's SequentialLR, which resets the LR to initial_lr and redoes
+        # the initial step of the first scheduler only.
+        self._apply_initial_lr(self._schedulers[0].get_last_lr())
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/ttml/common/schedulers.py:299]: Fixed in [333c883](https://github.com/tenstorrent/tt-metal/pull/54320/commits/333c883ddc786643d459a23b7ee0ef2eb1c81c63).

File: tt-train/sources/ttml/ttml/common/schedulers.py [hunk 2]
```diff
@@ -60,13 +60,33 @@ def set_lr(self, lr: float):
...
         self._optimizer = optimizer
-        self._base_lr = optimizer.get_lr()
+        self._base_lr = _initial_lr(optimizer)
         self._last_step = 0
         self._last_lr = self._base_lr
 
+    def _apply_initial_lr(self, lr: float):
+        """Write the construction-time LR, mirroring PyTorch's initial step."""
+        self._optimizer.set_lr(lr)
+        self._last_lr = lr
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/ttml/common/schedulers.py:97]: Fixed in [333c883](https://github.com/tenstorrent/tt-metal/pull/54320/commits/333c883ddc786643d459a23b7ee0ef2eb1c81c63).

File: tt-train/sources/ttml/schedulers/linear_scheduler.cpp [hunk 3]
```diff
@@ -13,13 +13,26 @@ namespace ttml::schedulers {
...
+    TT_FATAL(
+        start_factor > 0.0F && start_factor <= 1.0F,
+        "start_factor = {} expected to be greater than 0 and less or equal to 1.",
+        start_factor);
+    TT_FATAL(end_factor >= 0.0F && end_factor <= 1.0F, "end_factor = {} expected to be between 0 and 1.", end_factor);
+
+    // Mirror PyTorch, which applies start_factor at construction: the LR used
+    // before the first step() is already base_lr * start_factor.
+    m_last_lr = m_base_lr * m_start_factor;
+    optimizer->set_lr(m_last_lr);
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/schedulers/linear_scheduler.cpp:35]: Fixed in [333c883](https://github.com/tenstorrent/tt-metal/pull/54320/commits/333c883ddc786643d459a23b7ee0ef2eb1c81c63).

File: tt-train/sources/ttml/schedulers/step_scheduler.cpp [hunk 4]
```diff
@@ -15,10 +15,14 @@ StepScheduler::StepScheduler(optimizers::OptimizerBase *optimizer, size_t step_s
...
     m_last_step(0),
-    m_base_lr(optimizer->get_lr()),
+    m_base_lr(optimizer->get_initial_lr()),
     m_last_lr(m_base_lr) {
     TT_FATAL(step_size > 0, "step_size = {} must be greater than zero.", step_size);
     TT_FATAL(gamma > 0.0f, "gamma = {} must be greater than zero.", gamma);
+
+    // Mirror PyTorch's construction-time initial step: at step 0 the factor is
+    // gamma^0 == 1, so the LR is (re)set to the base LR.
+    optimizer->set_lr(m_last_lr);
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/schedulers/step_scheduler.cpp:25]: Fixed in [333c883](https://github.com/tenstorrent/tt-metal/pull/54320/commits/333c883ddc786643d459a23b7ee0ef2eb1c81c63).

File: tt-train/sources/ttml/schedulers/lambda_scheduler.cpp [hunk 5]
```diff
@@ -11,8 +11,11 @@ LambdaScheduler::LambdaScheduler(optimizers::OptimizerBase *optimizer, std::func
     LRSchedulerBase(optimizer),
     m_lr_lambda(std::move(lr_lambda)),
     m_last_step(0),
-    m_base_lr(optimizer->get_lr()),
-    m_last_lr(optimizer->get_lr()) {
+    m_base_lr(optimizer->get_initial_lr()),
+    m_last_lr(m_base_lr) {
+    // Mirror PyTorch's LambdaLR, which applies lr_lambda(0) at construction.
+    m_last_lr = m_base_lr * m_lr_lambda(0);
+    optimizer->set_lr(m_last_lr);
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/schedulers/lambda_scheduler.cpp:18]: Fixed in [333c883](https://github.com/tenstorrent/tt-metal/pull/54320/commits/333c883ddc786643d459a23b7ee0ef2eb1c81c63).

File: tt-train/sources/ttml/schedulers/sequential_scheduler.cpp [hunk 6]
```diff
@@ -47,6 +47,14 @@ SequentialScheduler::SequentialScheduler(
...
         }
     }
+
+    // The children were constructed back-to-back on the same optimizer, so the
+    // optimizer currently holds the LAST child's construction-time LR. Only the
+    // first child is active; restore its initial LR. Mirrors PyTorch's
+    // SequentialLR, which resets the LR to initial_lr and redoes the initial
+    // step of the first scheduler only.
+    m_last_lr = m_schedulers.front()->get_last_lr();
+    optimizer->set_lr(m_last_lr);
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/schedulers/sequential_scheduler.cpp:57]: Fixed in [333c883](https://github.com/tenstorrent/tt-metal/pull/54320/commits/333c883ddc786643d459a23b7ee0ef2eb1c81c63).

File: tt-train/sources/ttml/schedulers/cosine_annealing_scheduler.cpp [hunk 7]
```diff
@@ -14,10 +14,14 @@ CosineAnnealingScheduler::CosineAnnealingScheduler(optimizers::OptimizerBase* op
...
-    m_base_lr(optimizer->get_lr()),
+    m_base_lr(optimizer->get_initial_lr()),
     m_last_step(0),
-    m_last_lr(optimizer->get_lr()) {
+    m_last_lr(m_base_lr) {
     TT_FATAL(T_max > 0, "T_max = {} must be greater than zero.", T_max);
+
+    // Mirror PyTorch's construction-time initial step: at step 0 the cosine
+    // factor is 1, so the LR is (re)set to the base LR.
+    optimizer->set_lr(m_last_lr);
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/schedulers/cosine_annealing_scheduler.cpp:24]: Fixed in [333c883](https://github.com/tenstorrent/tt-metal/pull/54320/commits/333c883ddc786643d459a23b7ee0ef2eb1c81c63).

File: tt-train/sources/ttml/optimizers/optimizer_base.cpp [hunk 8]
```diff
@@ -12,6 +12,22 @@ namespace ttml::optimizers {
...
+}
+
+void OptimizerBase::save_initial_lr(serialization::StateDict& dict) const {
+    // If no scheduler has recorded a base LR yet, the current LR is the base.
+    dict["initial_lr"] = m_initial_lr.value_or(get_lr());
+}
+
+void OptimizerBase::restore_initial_lr(const serialization::StateDict& dict) {
+    m_initial_lr = serialization::get_value_type<float>(dict, "initial_lr");
+}
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/optimizers/optimizer_base.cpp:29]: This change is meant to match PyTorch behavior, and I would prefer to correct it early on before widespread checkpointing occurs. PyTorch doesn't maintain such legacy support logic in its checkpointing, and keeping legacy support logic will get messy. This choice was intentional.

[the remark this answers is not in the document]
Comment (jmalone-tt) [reply tt-train/sources/ttml/optimizers/optimizer_base.cpp:36]: Agreed - should we show a warning or error if `initial_lr` isn't present in the old checkpoint though?

**Review by jmalone-tt — APPROVED** [review]
Please address the comment about `initial_lr` potentially missing from old checkpoints, then good to merge

File: tt-train/sources/ttml/optimizers/optimizer_base.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/optimizers/optimizer_base.cpp:36]: Graceful error message added in [e6d5a9d](https://github.com/tenstorrent/tt-metal/pull/54320/commits/e6d5a9d0d9440b19e427edb68ff97a4bd3a2ffa8). Now it fails with a runtime error explaining that `initial_lr` must be present in the checkpoint and that old checkpoints without it are not supported.
